### ISSUE TITLE
[BUG] Redis script writes to read-only Redises

### ISSUE THREAD

**Issue by TaaviE** [rspamd/rspamd#4347]
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Read about bug reporting in general: https://rspamd.com/doc/faq.html#how-to-report-bugs-found-in-rspamd
    * Enabled relevant debugging logs: https://rspamd.com/doc/faq.html#how-to-debug-some-module-in-rspamd 
    * Checked the FAQs about Core files in case of fatal crash: https://rspamd.com/doc/faq.html#how-to-figure-out-why-rspamd-process-crashed
    * Tried ASAN package and obtained the ASAN report (if possible): https://rspamd.com/doc/faq.html#asan-builds 
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=%E2%9C%93&q=is%3Aissue+user%3Arspamd
    * Checked that there is not already an experimental package or master branch

**Describe the bug**

Read-only and write-capable Redis servers are merged into a single list [here](https://github.com/rspamd/rspamd/blob/868b8fac116d11487b346110713283c04a865bfa/lualib/lua_redis.lua#L1116) but when they're used [here](https://github.com/rspamd/rspamd/blob/868b8fac116d11487b346110713283c04a865bfa/lualib/lua_redis.lua#L1164) this behaviour is likely incorrect when Neural module tries to write its scripts to Redises [here](https://github.com/rspamd/rspamd/blob/cbdb21960f2a885237c8e8179ac1d698ad9c57c2/lualib/plugins/neural.lua#L198).

This is demonstrated with logspam like:
```
2022-11-25T11:53:12.375294+00:00 RSPAMD-1.LOC-1.NAIDE.EE [mail.info] rspamd[4809]: <5555d4>; lua; lua_redis.lua:1163: uploaded redis script to read-only:6380 with id 1, sha: 7632c1a283516a86a80e123616157788a83e1241
2022-11-25T11:53:12.375445+00:00 RSPAMD-1.LOC-1.NAIDE.EE [mail.info] rspamd[4809]: <5555d4>; lua; lua_redis.lua:1163: uploaded redis script to write-capable:6380 with id 1, sha: 7632c1a283516a86a80e123616157788a83e1241
2022-11-25T11:53:12.375605+00:00 RSPAMD-1.LOC-1.NAIDE.EE [mail.info] rspamd[4809]: <5555d4>; lua; lua_redis.lua:1163: uploaded redis script to read-only:6380 with id 1, sha: 7632c1a283516a86a80e123616157788a83e1241
2022-11-25T11:53:12.375664+00:00 RSPAMD-1.LOC-1.NAIDE.EE [mail.info] rspamd[4809]: <5555d4>; lua; lua_redis.lua:1163: uploaded redis script to write-capable:6380 with id 1, sha: 7632c1a283516a86a80e123616157788a83e1241
2022-11-25T11:53:12.375901+00:00 RSPAMD-1.LOC-1.NAIDE.EE [mail.info] rspamd[4809]: <5555d4>; lua; lua_redis.lua:1163: uploaded redis script to write-capable:6380 with id 1, sha: 7632c1a283516a86a80e123616157788a83e1241
2022-11-25T11:53:12.375953+00:00 RSPAMD-1.LOC-1.NAIDE.EE [mail.info] rspamd[4809]: <5555d4>; lua; lua_redis.lua:1163: uploaded redis script to read-only:6380 with id 1, sha: 7632c1a283516a86a80e123616157788a83e1241
```

### Steps to Reproduce

1. Configure read-only and write-capable Redis servers for rspamd
2. Enable the Neural module
3. Neural module gets into a loop and fails to upload its script to a read-only Redis

**Expected behavior**
Read-only Redises are not used for writing scripts.

### Versions
3.4

**Comment by vstakhov** [comment]
That's a very good catch and I think you are right in your assumptions. Thank you for the report!

**Comment by vstakhov** [comment]
Or wait, both read and write servers should all be able to accept scripts, as scripts themselves are not read or write, it is a runtime parameter `is_write` as a same script can be both read-only or not. There is one issue with not loaded scripts, but I'm not sure we are talking about the same issue.

**Comment by TaaviE** [comment]
@vstakhov 
> Or wait, both read and write servers should all be able to accept scripts

Maybe in theory, but in practice it keeps on looping and uploading the script to the read-only redises. That also eventually causes those redises to fill up `/tmp`.

Specifying only one, write-capable redis solved the problem.

**Comment by vstakhov** [comment]
Again, scripts are uploaded to all servers. It is the intended and desired behaviour.

**Comment by TaaviE** [comment]
@vstakhov 
Yes, it might be intended and desired but it does not work appropriately. It consumes the entirety of `/tmp`, causes logspam and the `neural` module does not work.

**Comment by vstakhov** [comment]
How can it consume any space in `/tmp`? What do you see in the redis logs?

**Comment by TaaviE** [comment]
> How can it consume any space in /tmp? What do you see in the redis logs?

Because each upload attempt creates a temporary file containing that uploaded script.

> What do you see in the redis logs?

Nothing. Safe to assume there are no errors, there just seems to be some kind of misdetection about the need to upload a script again.

Considering how removing the read-only redis resolved the issue, it is practically solved for me personally.

**Comment by TaaviE** [comment]
```
2023-03-15T09:09:27.161653+00:00 RSPAMD-1.LOC-1.NAIDE.EE [mail.err] rspamd[43290]: <594556>; lua; mx_check.lua:88: redis_cache_cb received error: READONLY You can't write against a read only replica.
2023-03-15T09:09:27.163027+00:00 RSPAMD-1.LOC-1.NAIDE.EE [mail.err] rspamd[43290]: <594556>; lua; history_redis.lua:132: got error READONLY You can't write against a read only replica. when writing history row: no value
```
This seems to impact more than one module actually, @vstakhov.

**Comment by schiznik** [comment]
I'm seeing this issue (with 3.8.1 on debian bullseye) - we have a redis cluster with sentinels in front. 

 lua; history_redis.lua:139: got error READONLY You can't write against a read only replica. when writing history row: no value
lua; replies.lua:142: redis_set_cb error when writing data to x.x.x.x: READONLY You can't write against a read only replica
lua; common.lua:301: failed to save virus cache for OK -> "rs_clamav_020a2546332298516369d5e48713885f": READONLY You can't write against a read only replica.

It looks to me to be the same issue as https://github.com/rspamd/rspamd/issues/4016 where workers weren't using sentinel settings from redis.conf
